File: hackerrank/2015.09/2015.09.23 - 101 Hack Sep 2015/Palindrome.java
package taskdirectory;

import hieule.utils.io.InputReader;
import java.io.PrintWriter;
import java.util.HashMap;
import java.util.Map;

public class Palindrome {

    private Map<Long, Integer> prefix;

    public void solve(int testNumber, InputReader in, PrintWriter out) {
        char[] s = in.next().toCharArray();
        int[] freq = new int[26];
        prefix = new HashMap<Long, Integer>();
        prefix.put((long)0, 1);

        long cnt = 0;
        for (int i = 0; i < s.length; i++) {
            int pos = s[i] - 'a';
            freq[pos] ^= 1;
            long mask = getMask(freq);
            cnt += count(mask);
            if (!prefix.containsKey(mask)) prefix.put(mask, 1);
            else prefix.put(mask, prefix.get(mask) + 1);
        }

        long total = (long)s.length * (s.length + 1) / 2;
        long gcd = findGcd(cnt, total);
        while (gcd > 1) {
            cnt /= gcd;
            total /= gcd;
            gcd = findGcd(cnt, total);
        }
        out.println(cnt + "/" + total);
    }

    private long getMask(int[] freq) {
        long res = 0;
        for (int i = 0; i < 26; i++)
            if (freq[i] == 1) res += (1 << i);
        return res;
    }

    private int count(long mask) {
        int res = 0;
        if (prefix.containsKey(mask)) res += prefix.get(mask);
        for (int i = 0; i < 26; i++) {
            long alternate;
            if ((mask & (1 << i)) == 0) alternate = mask + (1 << i);
            else alternate = mask - (1 << i);
            if (prefix.containsKey(alternate)) res += prefix.get(alternate);
        }
//        System.out.println(res);
        return res;
    }

    private long findGcd(long a, long b) {
        while (b != 0) {
            long temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
}

File: other/Palindrome.cpp
//
//  Palindrome.cpp
//  
//
//  Created by Hieu Le on 9/23/15.
//
//

#include <stdio.h>
#include <iostream>
#include <string>
#include <unordered_map>
#include <assert.h>

using namespace std;

typedef long long int_64;

int_64 findGcd(int_64 a, int_64 b) {
    while (b) {
        int_64 temp = b;
        assert (b);
        b = a % b;
        a = temp;
    }
    return a;
}

int_64 getMask(int freq[]) {
    long res = 0;
    for (int i = 0; i < 26; i++) {
        if (freq[i]) res += (1 << i);
    }
    return res;
}

int count(int_64 mask, unordered_map<int_64, int> prefix) {
    int res = 0;
    if (prefix.find(mask) != prefix.end()) res += prefix[mask];
    for (int i = 0; i < 26; i++) {
        long complement;
        if (mask & (1 << i)) complement = mask - (1 << i);
        else complement = mask + (1 << i);
        if (prefix.find(complement) != prefix.end()) res += prefix[complement];
    }
    return res;
}

int main() {
    int t;
    cin >> t;
    while (t--) {
        string s;
        cin >> s;
        int freq[26];
        for (int i = 0; i < 26; i++) {
            freq[i] = 0;
        }
        unordered_map<int_64, int> prefix;
        prefix[(int_64)0] = 1;
        
        int_64 cnt = 0;
        string::iterator iter;
        for (iter = s.begin(); iter != s.end(); iter++) {
            int pos = *iter - 'a';
            freq[pos] ^= 1;
            int_64 mask = getMask(freq);
            cnt += count(mask, prefix);
            if (prefix.find(mask) == prefix.end()) prefix[mask] = 1;
            else prefix[mask]++;
        }
        
        int_64 total = (int_64)s.size() * (s.size() + 1) / 2;
        int_64 gcd = findGcd(cnt, total);
        while (gcd > 1) {
            assert (gcd);
            cnt /= gcd;
            total /= gcd;
            gcd = findGcd(cnt, total);
        }
        cout << cnt << "/" << total << endl;
    }
}